java/alizer-api/src/main/java/com/redhat/devtools/alizer/api/ComponentRecognizerImpl.java

```diff
@@ -0,0 +1,301 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc.
+ ******************************************************************************/
+package com.redhat.devtools.alizer.api;
+
+import com.redhat.devtools.alizer.api.utils.DocumentParser;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class ComponentRecognizerImpl extends Recognizer {
+
+    public ComponentRecognizerImpl(RecognizerBuilder builder) {
+        super(builder);
+    }
+
+    public List<Component> analyze(String path) throws IOException {
+        List<File> files = getFiles(Paths.get(path));
+        List<Component> components = detectComponents(files);
+
+        // it may happen that a language has no a specific configuration file (e.g opposite to JAVA -> pom.xml and Nodejs -> package.json)
+        // we then rely on the language recognizer
+        List<Path> directoriesPathsWithoutConfigFile = getDirectoriesPathsWithoutConfigFile(Paths.get(path), components);
+        components.addAll(getComponentsWithoutConfigFile(directoriesPathsWithoutConfigFile));
+        return components;
+    }
+
+    private List<Path> getDirectoriesPathsWithoutConfigFile(Path root, List<Component> components) throws IOException {
+        if (components.isEmpty()) {
+            return Collections.singletonList(root);
+        }
+
+        List<Path> directoriesWithoutConfigFile = new ArrayList<>();
+        List<Path> allDirectoriesFromRoot = Files.walk(root, Integer.MAX_VALUE).filter(Files::isDirectory).skip(1).collect(Collectors.toList());
```

Comment:
Optimization: instead of generating a list from the stream and then re scanning the list could be rewritten as:
`.skip(1).foreach(x ->)`